sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -65,6 +65,11 @@ private[sql] case class InMemoryTableScanExec(
     case EqualTo(l: Literal, a: AttributeReference) =>
       statsFor(a).lowerBound <= l && l <= statsFor(a).upperBound
 
+    case EqualNullSafe(a: AttributeReference, l: Literal) if !l.nullable =>
+      statsFor(a).lowerBound <= l && l <= statsFor(a).upperBound
+    case EqualNullSafe(l: Literal, a: AttributeReference) if !l.nullable =>
```

Comment:
Ah, actually it seems okay because when `Literal` is `null`, it seems the filter is failed to resolved as the `Literal` type is `NullType` and then not applied for filtering here. And when the statistics for `upperBound` and `lowerBound` are `null`, it also seems okay because it seems the generated predicate deals with nullability.